8150564: Migrate useful ExtendedRobot methods into awt.Robot

alisenchung opened the pull request:
Some useful methods in ExtendedRobot should be migrated into Robot itself so that ExtendedRobot can be removed in the future. The tests using ExtendedRobot for these migrated methods are changed to use only Robot (removing unnecessary building of ExtendedRobot).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8349593](https://bugs.openjdk.org/browse/JDK-8349593) to be approved

### Error
&nbsp;⚠️ OCA signatory status must be verified

### Issues
 * [JDK-8150564](https://bugs.openjdk.org/browse/JDK-8150564): Migrate useful ExtendedRobot methods into awt.Robot (**Enhancement** - P3)
 * [JDK-8349593](https://bugs.openjdk.org/browse/JDK-8349593): Migrate useful ExtendedRobot methods into awt.Robot (**CSR**)


### Reviewers
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) 🔄 Re-review required (review applies to [b478652f](https://git.openjdk.org/jdk/pull/22044/files/b478652fe06acfd24cbef10c5ce1f7863043338b))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22044/head:pull/22044` \
`$ git checkout pull/22044`

Update a local copy of the PR: \
`$ git checkout pull/22044` \
`$ git pull https://git.openjdk.org/jdk.git pull/22044/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22044`

View PR using the GUI difftool: \
`$ git pr show -t 22044`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22044.diff">https://git.openjdk.org/jdk/pull/22044.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22044#issuecomment-2471189378)
</details>

prrace commented:
Please resolve the merge conflict.

prrace commented on src/java.desktop/share/classes/java/awt/Robot.java:
```diff
@@ -775,4 +775,250 @@ public synchronized String toString() {
+     * @see     java.awt.event.MouseEvent
+     */
+    public void click(int buttons) {
+//        throw new RuntimeException("click");
```
get rid of the comments.
Also you need @since  tags on the new methods

prrace commented on src/java.desktop/share/classes/java/awt/Robot.java:
```diff
@@ -775,4 +775,250 @@ public synchronized String toString() {
+     * @see     #click(int)
+     */
+    public void click() {
+//        throw new RuntimeException("click");
```
same

prrace commented on src/java.desktop/share/classes/java/awt/Robot.java:
```diff
@@ -775,4 +775,250 @@ public synchronized String toString() {
+     *          dispatching thread
+     */
+    public synchronized void waitForIdle(int delayValue) {
+//        throw new RuntimeException("waitForIdle");
```
same .. I'm not going to point them all out, just fix them

prrace commented:
/csr

prrace commented:
/reviewers reviewers 2

prrace commented:
/reviewers 2 reviewers

kumarabhi006 commented on test/jdk/java/awt/Frame/DecoratedExceptions/DecoratedExceptions.java:
Expand wild imports here and in other tests wherever applicable.

alisenchung replied:
out of scope of this change, i won't be changing any tests here outside of stuff relating to ExtendedRobot (also expanding imports of 200+ tests would be insane)

alisenchung commented:
/open

azvegint commented on test/jdk/lib/client/ExtendedRobot.java:
```diff
@@ -106,180 +103,8 @@ public ExtendedRobot(GraphicsDevice screen) throws AWTException {
-     */
-    @Override
-    public synchronized void waitForIdle() {
-        waitForIdle(syncDelay);
```
This `waitForIdle(500)` is no longer called by tests(as it uses regular `java.awt.Robot#waitForIdle()`, so I assume you have verified that automated testing looks good. (I haven't gone through all the tests yet)

azvegint commented on src/java.desktop/share/classes/java/awt/Robot.java:
```diff
@@ -126,6 +126,17 @@ public class Robot {
+    /**
+     * Default 2 pixel step length for mouse {@code glide}.
+     */
+    public static final int DEFAULT_STEP_LENGTH = 2;
```
Do we want to make the `DEFAULT_DELAY` and `DEFAULT_STEP_LENGTH` configurable?

prrace replied:
glide  has over-rides step length

alisenchung replied:
I think setting the default step length with a setter/getter makes sense. I think generally tests don't change the step delays across their glides anyway and it would remove the need for all those extra overrides for glide. I'll make this change now

azvegint commented on src/java.desktop/share/classes/java/awt/Robot.java:
```diff
@@ -773,4 +784,175 @@ public synchronized String toString() {
+             stepNum++) {
+            x += tDx;
+            y += tDy;
+            mouseMove((int)x, (int)y);
```
`mouseMove` can throw an `IllegalThreadStateException` under certain circumstances.

prrace replied:
yes should be added to all glide API methods

azvegint commented on src/java.desktop/share/classes/java/awt/Robot.java:
```diff
@@ -773,4 +784,175 @@ public synchronized String toString() {
+     * @see     #glide(int, int, int, int, int, int)
+     * @since   25
+     */
+    public void glide(int x, int y) {
```
I don't see the `public void glide(Point dest)` and `public void glide(Point src, Point dest)` added, it may be convenient in some cases.

alisenchung replied:
I've taken a look at all the tests running glide and very few of them actually used glide(Point dest) and glide(Point src, Point dest) so I decided to remove them from them from the migration

azvegint commented on test/jdk/java/awt/Component/PaintAll/PaintAll.java:
```diff
@@ -48,7 +49,6 @@
   @summary Test Component.paintAll() method
   @author sergey.bylokhov@oracle.com: area=awt.component
   @library /lib/client/
```
It's not needed anymore, is it?
```suggestion
```

azvegint commented on test/jdk/lib/client/ExtendedRobot.java:
```diff
@@ -334,62 +158,4 @@ public void dragAndDrop(int fromX, int fromY, int toX, int toY){
-     * @see     #type(int)
-     * @see     java.awt.event.KeyEvent
-     */
-    public void type(char c) {
```
Those `type(char...` are not migrated also. 

At least one test uses it:

https://github.com/alisenchung/jdk/blob/8150564/test/jdk/java/awt/Window/ShapedAndTranslucentWindows/ShapedTranslucentWindowClick.java#L178

With your change, it now calls `type(int)` directly, which has a different implementation.

alisenchung commented:
/open

prrace commented on src/java.desktop/share/classes/java/awt/Robot.java:
```diff
@@ -126,6 +126,17 @@ public class Robot {
     private DirectColorModel screenCapCM = null;
 
+    /**
+     * Default 20 milliseconds delay for mouse {@code click} and
```
Be aware that values of static fields are automatically documented.
eg see
(https://docs.oracle.com/en/java/javase/21/docs/api/constant-values.html#java.awt.Font.BOLD)

And an app really could use knowing what the default is.
And these defaults have been battle-tested in ExtendedRobot

prrace commented on src/java.desktop/share/classes/java/awt/Robot.java:
```diff
@@ -773,4 +784,175 @@ public synchronized String toString() {
     }
+
+    /**
+     * A convenience method that simulates clicking a mouse button by calling {@code mousePress}
+     * and {@code mouseRelease}. Invokes {@code waitForIdle} with a default {@link #DEFAULT_DELAY delay} after
+     * {@code mousePress} and {@code mouseRelease} calls. For specifics on valid inputs please see
+     * {@link java.awt.Robot#mousePress(int)}.
```
I understand the concern but in this case it is more or less explanation of what the method does. Perhaps it can be adjusted or made in to an apiNote or whatever is best.

honkar-jdk commented on test/jdk/lib/client/ExtendedRobot.java:
```diff
@@ -106,180 +103,8 @@ public ExtendedRobot(GraphicsDevice screen) throws AWTException {
-     *
-     * @see     #glide(int, int, int, int, int, int)
-     */
-    public void glide(Point src, Point dest) {
-        glide(src.x, src.y, dest.x, dest.y, DEFAULT_STEP_LENGTH, DEFAULT_SPEED);
-    }
```
Shouldn't `glide(Point src, Point dest)` and `glide(Point dest)` be retained in ExtendedRobot
In a [previous comment](https://github.com/openjdk/jdk/pull/22044/files#r2085671341) it was mentioned that these methods won't be migrated so do they need to be retained in ExtendedRobot ?

I see `MultipleMouseButtonsTest.java` using `glide(Point src, Point dest)`. There are two options here:

1) Either update the test to use Robot's glide()
change the line `robot.glide(origin, center)` to `robot.glide(origin.x, origin.y, center.x, center.y);`  and completely remove ExtendedRobot versions of glide.

2) Or retain these two convenience methods in ExtendedRobot.

honkar-jdk reviewed (commented):
I see couple of tests that use ExtendedRobot which can be changed to use Robot. Can you double check these tests.

```
java/awt/Frame/DisposeParentGC/DisposeParentGC.java
java/awt/List/SetBackgroundTest/SetBackgroundTest.java
java/awt/Modal/ToBack/ToBackFDFTest.java
```

honkar-jdk commented on test/jdk/java/awt/TrayIcon/TrayIconPopup/TrayIconPopupTest.java:
```diff
@@ -53,7 +54,6 @@
  *          java.desktop/java.awt.Helper
  *          jdk.test.lib.Platform
  *          jtreg.SkippedException
- *          ExtendedRobot
```
Since ExtendedRobot is removed /lib/client is no longer required in `@library` tag. Applicable for other tests as well.

honkar-jdk replied:
There are couple of tests with /lib/client in `@library` tag which is no longer required. This cleanup can be handled in subsequent PR.

honkar-jdk reviewed (changes requested):
Since there are many test changes it is good to run CI on latest update.

honkar-jdk reviewed (commented):
Latest changes LGTM. 

@alisenchung 
Does CI run look good with the latest changes?

alisenchung commented:
I've decided to remove click/type's usage of the default step-delay and just forced them to use a delay of 20ms since I don't think anyone would need to explicitly set a delay for clicking and typing. I also made the step delay/length to be configurable for glide and added setters/getters for them. Because of that I removed the last glide override that explicitly set the delay and length.

prrace commented on test/jdk/lib/client/ExtendedRobot.java:
```diff
@@ -106,180 +103,8 @@ public ExtendedRobot(GraphicsDevice screen) throws AWTException {
-     * @see     #mouseMove(int, int)
-     * @see     #delay(int)
-     */
-     public void glide(int srcX, int srcY, int destX, int destY, int stepLength, int speed) {
```
I hadn't noticed that this over-ride is both removed here and not added to Robot.
If this were available, then perhaps those setters and getters for the current values aren't needed !
What was the reason for not including it ? No one ever used it or something else ?

prrace commented on src/java.desktop/share/classes/java/awt/Robot.java:
```diff
@@ -773,4 +786,238 @@ public synchronized String toString() {
+     * @param   destX  Destination point x coordinate
+     * @param   destY  Destination point y coordinate
+     *
+     * @throws  IllegalArgumentException if {@code stepLength} is greater than the distance
```
In the CSR Joe asked if glide(int, int) needed to throw IAE and it made me look here and I don't actually see where this code would throw that. 
And the old extended robot never had such a feature, did it ?
Also if you think of glide as a sequence of steps, then even with the default of "2" if the distance to move is say "5" after 2 steps you are one pixel away and then I presume we just step one pixel for the final step.
So how is that different than STARTING one pixel away ?

Perhaps we can just dispense with the IAE and if dist < steplength then there's just one step.

prrace commented on src/java.desktop/share/classes/java/awt/Robot.java:
```diff
@@ -773,4 +783,210 @@ public synchronized String toString() {
+
+    /**
+     * A convenience method that moves the mouse in multiple
+     * steps from its current location to the destination coordinates. Invokes
```
I think this is the text that caused Joe to write
"From reading the text, it should like the method would benefit from an @implSpec tag describing the behavior."
You can read about implSpec here https://bugs.openjdk.org/browse/JDK-8068562

azvegint commented on test/jdk/lib/client/ExtendedRobot.java:
```diff
@@ -334,62 +158,4 @@ public void dragAndDrop(int fromX, int fromY, int toX, int toY){
     public void dragAndDrop(Point from, Point to){
```
I missed that earlier. Why didn't we also move the `dragAndDrop` methods to `java.awt.Robot`?
This would allow us to remove the ExtendedRobot class entirely.

azvegint replied:
>  i found that dragAndDrop was never used by any tests

This doesn't seem to be true. I see several tests using it:

```
./java/awt/Window/ShapedAndTranslucentWindows/StaticallyShaped.java:73:        robot.dragAndDrop(location.x + dl, location.y + 5, location.x + dl + random.nextInt(dl), location.y + random.nextInt(dl));
./java/awt/Window/ShapedAndTranslucentWindows/StaticallyShaped.java:79:        robot.dragAndDrop(location.x + 4, location.y + 4, location.x + random.nextInt(2*dl)-dl, location.y + random.nextInt(2*dl)-dl);
./java/awt/Window/ShapedAndTranslucentWindows/ShapedTranslucent.java:75:        robot.dragAndDrop(location.x + dl, location.y + 5, location.x + dl + random.nextInt(dl), location.y + random.nextInt(dl));
./java/awt/Window/ShapedAndTranslucentWindows/ShapedTranslucent.java:81:        robot.dragAndDrop(location.x + 4, location.y + 4, location.x + random.nextInt(2*dl)-dl, location.y + random.nextInt(2*dl)-dl);
./java/awt/Window/ShapedAndTranslucentWindows/Translucent.java:72:        robot.dragAndDrop(location.x + dl, location.y + 5, location.x + dl + random.nextInt(dl), location.y + random.nextInt(dl));
./java/awt/Window/ShapedAndTranslucentWindows/Translucent.java:78:        robot.dragAndDrop(location.x + 4, location.y + 4, location.x + random.nextInt(2*dl)-dl, location.y + random.nextInt(2*dl)-dl);
./java/awt/Window/ShapedAndTranslucentWindows/Shaped.java:75:        robot.dragAndDrop(location.x + dl, location.y + 5, location.x + dl + random.nextInt(dl), location.y + random.nextInt(dl));
./java/awt/Window/ShapedAndTranslucentWindows/Shaped.java:81:        robot.dragAndDrop(location.x + 4, location.y + 4, location.x + random.nextInt(2*dl)-dl, location.y + random.nextInt(2*dl)-dl);
./java/awt/Window/ShapedAndTranslucentWindows/ShapedTranslucentWindowClick.java:174:        robot.dragAndDrop(MouseInfo.getPointerInfo().getLocation(), new Point(x+5, y));
./java/awt/event/MouseEvent/DragToLightweightTest.java:79:        robot.dragAndDrop(
./javax/swing/JWindow/ShapedAndTranslucentWindows/Common.java:369:        robot.dragAndDrop(location.x + 30, location.y + 5, location.x + dl + random.nextInt(dl), location.y + random.nextInt(dl));
./javax/swing/JWindow/ShapedAndTranslucentWindows/Common.java:375:        robot.dragAndDrop(location.x + 4, location.y + 4, location.x + random.nextInt(2*dl)-dl, location.y + random.nextInt(2*dl)-dl);
./javax/swing/JWindow/ShapedAndTranslucentWindows/Common.java:387:        robot.dragAndDrop(location.x + 30, location.y + 5, location.x + dl + random.nextInt(dl), location.y + random.nextInt(dl));
./javax/swing/JWindow/ShapedAndTranslucentWindows/Common.java:393:        robot.dragAndDrop(location.x + 4, location.y + 4, location.x + random.nextInt(2*dl)-dl, location.y + random.nextInt(2*dl)-dl);
```

It seems to be a convenient method because it eliminates the need for all this hassle with a single call.


```java
robot.mouseMove(fromX, fromY);
robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
robot.waitForIdle();
robot.glide(toX, toY);
robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
robot.waitForIdle();
```

vs

```java
robot.dragAndDrop(fromX, fromY, toX, toY)
```

alisenchung replied:
I've migrated dragAndDrop(int, int, int, int) and updated the csr to include the migration. Can you take a look at the specification to see if it meets the standard?

azvegint commented on src/java.desktop/share/classes/java/awt/Robot.java:
```diff
@@ -773,4 +783,241 @@ public synchronized String toString() {
+    /**
+     * A convenience method that drags the mouse with mouse button 1 pressed
+     * from source point to destination point.
+     *
+     * @implSpec The method moves the mouse pointer to source point ({@code fromX}, {@code fromY}),
+     * presses mouse button 1, calls glide to drag pointer to
+     * destination point ({@code toX}, {@code toY}), then releases mouse button 1.
```
It can probably be omitted because it's essentially the same information that's already stated in the previous sentence and the `@param` tags.

azvegint commented on test/jdk/java/awt/datatransfer/Independence/IndependenceSwingTest.java:
```diff
@@ -27,7 +27,6 @@
  * @requires (os.family == "linux")
  * @summary To make sure that System & Primary clipboards should behave independently
  * @library /lib/client
```
> I've updated the current tests in to remove /client/lib from the tests I already removed ER from

However, it seems that some of the touch tests are still using the ER, which will result in a compilation failure when run with CI.
For instance, the `IndependenceSwingTest.java` test still fails to compile and run. There are more examples.

> I'll remove the remaining tests when I remove ER in a followup PR

Sure, thank you.
